File: leetcode/JavaScript/No215.kth-largest-element-in-an-array.js
/**
 * Difficulty:
 * Medium
 *
 * Desc:
 * Find the kth largest element in an unsorted array.
 * Note that it is the kth largest element in the sorted order, not the kth distinct element.
 *
 * Example:
 * Given [3,2,1,5,6,4] and k = 2, return 5.
 *
 * Note:
 * You may assume k is always valid, 1 ≤ k ≤ array's length.
 *
 * 求乱序数组中第 K 大的元素。最小堆问题
 */

/**
 * ========================================= Solution 1 =========================================
 * 最小堆
 */

const buildFromTail = (array) => {
  let binaryHeaps = [...array];
  for (let i = binaryHeaps.length - 1; i >= 0; i -= 1) {
    // 每一个元素作为父元素，和其子元素进行比较
    binaryHeaps = sortWithChild(binaryHeaps, i + 1);
  }
  return binaryHeaps;
};

const less = (num1, num2) => num1 < num2;
const exchange = (arr, i1, i2) => {
  const tmp = arr[i1];
  arr[i1] = arr[i2];
  arr[i2] = tmp;
  return arr;
};

const sortWithChild = (heaps, fatherIndex) => {
  const childIndexLeft = fatherIndex * 2;
  const childIndexRight = childIndexLeft + 1;

  if (childIndexLeft > heaps.length) return heaps;
  // 如果子元素的大小 < 父元素，则交换位置
  if (less(heaps[childIndexLeft - 1], heaps[fatherIndex - 1])) {
    heaps = exchange(heaps, childIndexLeft - 1, fatherIndex - 1);
  }

  if (childIndexRight > heaps.length) return heaps;
  if (less(heaps[childIndexRight - 1], heaps[fatherIndex - 1])) {
    heaps = exchange(heaps, childIndexRight - 1, fatherIndex - 1);
  }

  // 将两个子元素作为父元素，继续和它们的子元素进行排序
  heaps = sortWithChild(heaps, childIndexLeft);
  heaps = sortWithChild(heaps, childIndexRight);
  return heaps;
};

/**
* @param {number[]} nums
* @param {number} k
* @return {number}
*/
var findKthLargest_1 = function(nums, k) {
  const kItems = nums.slice(0, k);
  const heaps = buildFromTail(kItems);

  for (let i = k; i < nums.length; i += 1) {
    const val = nums[i];
    // 如果元素大于堆顶，则交换值，并重新排列堆
    if (val > heaps[0]) {
      heaps[0] = val;
      sortWithChild(heaps, 1);
    }
  }
  return heaps[0];
};

/**
 * ========================================= Solution 2 =========================================
 * 3-way-partion 三向切分
 */

const partition = (nums, i, k, target) => {
  if (i >= k) return i

  const swap = (i1, i2) => {
    const tmp = nums[i1]
    nums[i1] = nums[i2]
    nums[i2] = tmp
  }

  let j = i
  while (j <= k) {
    if (nums[j] > target) {
      swap(j, k)
      k -= 1
    } else if (nums[j] < target) {
      swap(i, j)
      i += 1
      j += 1
    } else {
      j += 1
    }
  }

  return i
}


/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 *
 * 3-way-partion 三向切分
 */
var findKthLargest_2 = function(nums, k) {
  let i = 0
  let j = nums.length

  const getRandomIndex = (i1, i2) => Math.floor(Math.random() * (i2 - i1) + i1)

  while (true) {
    const mid = getRandomIndex(i, j)
    const pivot = partition(nums, i, j - 1, nums[mid])
    if (pivot + k === nums.length) return nums[pivot]
    if (pivot + k < nums.length) {
      i = pivot + 1
    } else {
      j = pivot
    }
  }
}

File: leetcode/C#/No215.kth-largest-element-in-an-array.cs
/*
 * Difficulty:
 * Medium
 *
 * Desc:
 * Find the kth largest element in an unsorted array.
 * Note that it is the kth largest element in the sorted order, not the kth distinct element.
 *
 * Example 1:
 * Input: [3,2,1,5,6,4] and k = 2
 * Output: 5
 *
 * Example 2:
 * Input: [3,2,3,1,2,4,5,5,6] and k = 4
 * Output: 4
 *
 * Note:
 * You may assume k is always valid, 1 ≤ k ≤ array's length.
*/

public interface IBinaryHeapEntity
{
    public int Value { get; }
}

public interface IBinaryHeap<T>
{
    public int Count { get; }

    public void Push(T item);

    public T Pop();

    public T Top();

    public void Exchange(T item, int pos);
}

public class BinaryHeapEntity : IBinaryHeapEntity
{
    private int Val;

    public BinaryHeapEntity(int value)
    {
        this.Val = value;
    }

    public int Value
    {
        get
        {
            return this.Val;
        }
    }
}

// 最小堆
class BinaryHeap<T> : IBinaryHeap<T>
        where T : IBinaryHeapEntity
{
    private readonly List<T> heap;

    public BinaryHeap()
    {
        this.heap = new List<T>();
    }

    public int Count
    {
        get
        {
            return this.heap.Count;
        }
    }

    public void Push(T entity)
    {
        this.heap.Add(entity);
        this.SortWithFather(this.Count);
    }

    public T Top()
    {
        if (this.Count == 0)
        {
            return default;
        }

        return this.heap[0];
    }

    public T Pop()
    {
        if (this.Count == 0)
        {
            return default;
        }

        T lastEntity = this.heap[this.Count - 1];
        this.heap[this.Count - 1] = this.heap[0];
        this.heap[0] = lastEntity;

        T res = this.heap[this.Count - 1];
        this.heap.RemoveAt(this.Count - 1);
        if (this.Count > 0)
        {
            this.SortWithChildren(1);
        }

        return res;
    }

    public void Exchange(T entity, int pos)
    {
        this.heap[pos] = entity;
        this.SortWithChildren(pos + 1);
    }

    private int GetValue(int index)
    {
        return this.heap[index].Value;
    }

    private void SortWithFather(int cPos)
    {
        int fPos = cPos / 2;
        T target = this.heap[cPos - 1];

        while (cPos > 1 && target.Value < this.GetValue(fPos - 1))
        {
            this.heap[cPos - 1] = this.heap[fPos - 1];
            cPos = fPos;
            fPos = cPos / 2;
        }

        this.heap[cPos - 1] = target;
    }

    private void SortWithChildren(int fPos)
    {
        int cPos = fPos * 2;
        T target = this.heap[fPos - 1];

        while (cPos - 1 < this.Count)
        {
            if (cPos < this.Count && this.GetValue(cPos - 1) > this.GetValue(cPos))
            {
                cPos += 1;
            }

            if (this.GetValue(cPos - 1) >= target.Value)
            {
                break;
            }

            this.heap[fPos - 1] = this.heap[cPos - 1];
            fPos = cPos;
            cPos = fPos * 2;
        }

        this.heap[fPos - 1] = target;
    }
}

public class Solution {
    public int FindKthLargest(int[] nums, int k) {
        IBinaryHeap<IBinaryHeapEntity> heap = new BinaryHeap<IBinaryHeapEntity>();

        for (int i = 0; i < k; i += 1) {
            heap.Push(new BinaryHeapEntity(nums[i]));
        }

        for (int i = k; i < nums.Length; i += 1) {
            if (nums[i] > heap.Top().Value) {
                heap.Exchange(new BinaryHeapEntity(nums[i]), 0);
            }
        }
        return heap.Top().Value;
    }
}